README.md
<h1 align="center">
  mini-sync
</h1>
<p align="center">
  <a href="https://www.npmjs.org/package/mini-sync"><img src="https://badgen.net/npm/v/mini-sync" alt="npm"></a>
  <a href="https://github.com/rdmurphy/mini-sync/actions?query=workflow%3A%22Node.js+CI%22"><img src="https://badgen.net/github/checks/rdmurphy/mini-sync?label=build" alt="build status"></a>
  <a href="https://unpkg.com/mini-sync/client/dist/client.js"><img src="https://badgen.net/badgesize/gzip/https://unpkg.com/mini-sync/client/dist/client.js" alt="gzip size"></a>
  <a href="https://unpkg.com/mini-sync/client/dist/client.js"><img src="https://badgen.net/badgesize/brotli/https://unpkg.com/mini-sync/client/dist/client.js" alt="brotli size"></a>
  <a href="https://packagephobia.now.sh/result?p=mini-sync"><img src="https://badgen.net/packagephobia/install/mini-sync" alt="install size"></a>
</p>

`mini-sync` is an incredibly tiny, live-reloading development server that uses [server-sent events](https://developer.mozilla.org/en-US/docs/Web/API/Server-sent_events) to keep your browser in sync with your front-end code.

## Key features

- 🦗 For less **than a few hundred KBs**, get a fully functional static server that can communicate with browsers during development
- ♻️ Bundles the maintained version of [`livereload-js`](https://github.com/livereload/livereload-js) in the client code to manage the reloading logic
- 📦 Client-side code may be **automatically injected into HTML files** or included in browsers via your preferred bundler, the static server or CDN

## Table of contents

<!-- START doctoc generated TOC please keep comment here to allow auto update -->

<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->

<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->

- [Setup](#setup)
- [Usage](#usage)
  - [Server](#server)
  - [Client](#client)
    - [Automatically inject the script into HTML](#automatically-inject-the-script-into-html)
    - [Load directly in your HTML](#load-directly-in-your-html)
    - [Conditionally add it to your bundle](#conditionally-add-it-to-your-bundle)
- [API](#api)
  - [Table of Contents](#table-of-contents)
  - [CreateReturn](#createreturn)
    - [Properties](#properties)
  - [StartReturn](#startreturn)
    - [Properties](#properties-1)
  - [create](#create)
    - [Parameters](#parameters)
    - [Examples](#examples)
  - [reload](#reload)
    - [Parameters](#parameters-1)
  - [close](#close)
  - [start](#start)
- [Possible future features](#possible-future-features)
- [License](#license)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

## Setup

`mini-sync` has two layers - the server code that's responsible for serving static assets and delivering reload requests, and the client code that needs to be present on your page to receive messages from the server.

Install the package with `npm`, `yarn`, or `pnpm`:

```sh
npm install --save-dev mini-sync
# or
yarn add --dev mini-sync
# or
pnpm add --save-dev mini-sync
```

## Usage

You will need to integrate `mini-sync` both into your build pipeline and your JavaScript/HTML.

### Server

Implement `mini-sync` in your build tool by using it as the static server for your assets during development. Once the server is created, it will return a `reload` function that can be called any time you need to communicate with the browser, a `start` function for activating the static server and watching for `reload` calls, and a `close` function for stopping the server.

```js
const chokidar = require('chokidar'); // or your preferred file watcher
const { create } = require('mini-sync');

const dirToWatch = './app';

async function main() {
  const server = create({
    dir: dirToWatch,
    port: 3000,
  });

  const watcher = chokidar.watch('.', { cwd: dirToWatch });

  // every time a file changes, we call the reload command
  watcher.on('change', (path) => {
    server.reload(path);
  });

  // start our dev server
  const { local, network } = await server.start();

  // report out what our URLs are
  console.log(`Local server URL: ${local}`); // http://localhost:3000
  console.log(`Local network URL: ${network}`); // http://127.x.x.x:3000

  // ...some time later
  await server.close();
}

main().catch(console.error);
```

### Client

`mini-sync` has a tiny script that needs to be added to your JavaScript bundles or loaded on your HTML page. How best to go about this will depend on your environment, but there are a few methods to consider.

#### Automatically inject the script into HTML

As of 0.4.0 the easiest method for adding the client-side script is to let `mini-sync` automatically inject the import whenever it serves HTML. As long as the HTML has a closing `</head>` tag, the `mini-sync` server can append the script tag to the `<head>`.

This happens **by default** as of 0.4.0. If you'd like to disable this feature, pass `false` to `injectClientScript` when creating a `mini-sync` instance.

```js
const server = create({
  dir: dirToWatch,
  injectClientScript: false, // defaults to "true"
  port: 3000,
});
```

#### Load directly in your HTML

You can also add it to your HTML manually. Ideally it would run _before_ the rest of your JavaScript does. If you plan to use this method, you should set `injectClientScript` to `false` when creating the `mini-sync` instance so the auto-injection does not interfere with your manually added copy.

As of 0.2.0 the `mini-sync` server hosts its own copy of the client script, and it can be referenced in your HTML.

```html
<script async src="/__mini_sync__/client.js"></script>
```

It's also possible to load it via [unpkg.com](https://unpkg.com/), but this is not recommended due to the risk of version mismatch.

```html
<script async src="https://unpkg.com/mini-sync"></script>
```

#### Conditionally add it to your bundle

Most bundlers support conditional includes based on the value of the `NODE_ENV` environment variable, or a similar mechanism. If you can do this in the configuration itself, that's great! But you also could include it directly in your JavaScript itself within an `if` statement. If you plan to use this method, you should set `injectClientScript` to `false` when creating the `mini-sync` instance so the auto-injection does not interfere with your bundled copy.

```js
if (process.env.NODE_ENV === 'development') {
  require('mini-sync/client');
}
```

In this case it will be present in development builds, but in production builds it will be skipped or entirely removed by your minifier.

## API

<!-- Generated by documentation.js. Update this documentation by updating the source code. -->

#### Table of Contents

- [CreateReturn](#createreturn)
  - [Properties](#properties)
- [StartReturn](#startreturn)
  - [Properties](#properties-1)
- [create](#create)
  - [Parameters](#parameters)
  - [Examples](#examples)
- [reload](#reload)
  - [Parameters](#parameters-1)
- [close](#close)
- [start](#start)

### CreateReturn

What's returned when the `create` function is called.

Type: [object](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Object)

#### Properties

- `close` **[Function](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Statements/function)** Stops the server if it is running
- `reload` **[Function](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Statements/function)** When called this function will reload any connected HTML documents, can accept the path to a file to target for reload
- `start` **[Function](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Statements/function)** When called the server will begin running

### StartReturn

What's returned by the `start` function in a Promise.

Type: [object](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Object)

#### Properties

- `local` **[string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** The localhost URL for the static site
- `network` **[string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** The local networked URL for the static site
- `port` **[number](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Number)** The port the server ended up on

### create

Creates a server on the preferred port and begins serving the provided
directories locally.

#### Parameters

- `options` **[object](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Object)** (optional, default `{}`)
  - `options.dir` **([string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String) \| [Array](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Array)&lt;[string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)>)?** The directory or list of directories to serve (optional, default `process.cwd()`)
  - `options.injectClientScript` **[boolean](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Boolean)?** If true, inject the client script into served HTML pages (optional, default `true`)
  - `options.port` **[number](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Number)?** The port to serve on (optional, default `3000`)

#### Examples

```javascript
const { create } = require('mini-sync');

const server = create({ dir: 'app', port: 3000 });

const { local } = await server.start();

console.log(`Now serving at: ${local}`);

// any time a file needs to change, use "reload"
server.reload('app.css');

// reloads the whole page
server.reload();

// close the server
await server.close();
```

Returns **[CreateReturn](#createreturn)**

### reload

Tells all connected clients to reload.

#### Parameters

- `file` **[string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)?** The file to reload

### close

Returns a promise once the server closes.

Returns **[Promise](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Promise)&lt;void>**

### start

Returns a promise once the server starts.

Returns **[Promise](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Promise)&lt;[StartReturn](#startreturn)>**

## Possible future features

- The ability to additionally proxy existing servers

## License

MIT


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

### Added

- A custom favicon for pages being served by `mini-sync` is now available.
- It is now possible for `mini-sync` to automatically inject a `<script>` tag adding the served copy at `__mini_sync__/client.js` to HTML pages. This is configured to **happen by default** and can be disabled by setting `injectClientScript` to `false` when creating a `mini-sync` instance.

## [0.3.0] - 2020-08-11

### Changed

- No code changes, but via dependency updates (in `sirv`) all files served by `mini-sync` will now receive a `no-cache` Cache Control header.

## [0.2.0] - 2020-05-13

### Added

- `mini-sync` will now try to recover if the provided port is taken (defaults to `3000`) and increment until it successfully finds an open one.

- The `mini-sync` server now serves a copy of the client side script at `__mini_sync__/client.js`. This may be useful if a user wants to include the client side script without having to depend on `unpkg.com`.

### Changed

- The exported `create` function now returns a `reload` function (for setting up reload calls) a `start` function (activates the static server and returns a Promise with the expected `local` and `network` URLs) and a `close` function. This makes it possible to prepare the server and pass around the `reload` function without the server being active until you're ready, and to spin it down manually. Previously it was possible for `mini-sync` to "beat" the rest of your code to being ready, causing cryptic race condition errors.

- An implementation detail, but `mini-sync`'s SSE URL is now `__mini_sync__` instead of `__dev_sync__` - an accidential leftover from a previous iteration of the library.

- The client code now logs `[MINI SYNC]` instead of `[DEV SYNC]` so the source of the client-side logging is clearer.

- A custom logger is now being passed to LiveReload so it won't be so chatty in the console. Instead, `mini-sync` has its own minimal logging of what the `reload` function sent to the client.

### Fixed

- You were technically suppose to be able to not provide any options to `create` but previously it would throw an error. Now passing no options is valid as originally intended.

## [0.1.0] - 2019-11-03

### Added

- Initial release!


test/test.node.js
// native
const assert = require('assert').strict;
const fs = require('fs').promises;
const http = require('http');
const { join } = require('path');

// packages
const fetch = require('node-fetch').default;
const { suite } = require('uvu');

// library
const { create } = require('../server');

const basic = suite('basic mini-sync serving tests');

let server;
let base;

basic.before(async () => {
  // create the server
  server = create({ dir: join(__dirname, 'fixtures/basic') });

  // start the server
  const { local } = await server.start();
  base = local;
});

basic.after(async () => {
  await server.close();
});

basic('should serve static files', async () => {
  const actual = await fetch(new URL('file.txt', base)).then((res) =>
    res.text()
  );

  assert.equal(actual, 'This is a served file.\n');
});

basic('should be possible to hit the event-stream endpoint', () => {
  const req = http.get(new URL('__mini_sync__', base), (res) => {
    req.abort();

    assert.ok(res.statusCode === 200);
  });
});

basic('should serve the client library', async () => {
  const actual = await fetch(
    new URL('__mini_sync__/client.js', base)
  ).then((res) => res.text());

  const expected = await fs.readFile(
    join(process.cwd(), 'client/dist/client.js'),
    'utf8'
  );

  assert.equal(actual, expected);
});

basic('should serve the favicon.ico', async () => {
  const actual = await fetch(new URL('favicon.ico', base)).then((res) =>
    res.text()
  );

  const expected = await fs.readFile(
    join(process.cwd(), 'assets/favicon.ico'),
    'utf8'
  );

  assert.equal(actual, expected);
});

basic.run();

const options = suite('mini-sync options');

options('should be possible to pass in a different port', async () => {
  const expectedPort = 1234;
  const expectedLocal = `http://localhost:${expectedPort}`;

  const server = create({ port: expectedPort });

  const { local, port } = await server.start();

  assert.equal(port, expectedPort);
  assert.equal(local, expectedLocal);

  await server.close();
});

options(
  'should attempt to find a new port if the provided one is taken',
  async () => {
    const takenPort = 5000;
    const blockingServer = http.createServer().listen(takenPort);

    const server = create({ port: takenPort });

    const { port } = await server.start();

    assert.equal(port, takenPort + 1);

    blockingServer.close();
    await server.close();
  }
);

options.run();


test/test.browser.js
// native
const assert = require('assert').strict;
const fs = require('fs').promises;
const { tmpdir } = require('os');
const { join } = require('path');

// packages
const { chromium, webkit, firefox } = require('playwright');
const { suite } = require('uvu');

// library
const { create } = require('../server');

const browserName = process.env.BROWSER || 'chromium';

const it = suite('browser tests');

let browser;
let page;
let server;
let url;
let eventSourceUrl;
let clientUrl;

it.before(async function () {
  server = create({ dir: join(__dirname, 'fixtures/basic') });
  const { local } = await server.start();

  url = local;
  eventSourceUrl = `${url}/__mini_sync__`;
  clientUrl = `${eventSourceUrl}/client.js`;

  browser = await { chromium, webkit, firefox }[browserName].launch();
  page = await browser.newPage();
});

it.after(async () => {
  await page.close();
  await browser.close();
  await server.close();
});

it('should load client script and connect using EventSource', async () => {
  // prep to track our responses
  const responses = new Set();

  page.on('response', (res) => {
    responses.add(res.url());
  });

  // load the page
  await page.goto(url);

  // wait for our event source to connect
  await page.waitForResponse(eventSourceUrl);

  // assert we loaded everything we expected
  assert.ok(responses.has(clientUrl));
  assert.ok(responses.has(eventSourceUrl));
});

it('should do a hard reload', async () => {
  // load the page
  await page.goto(url);

  // wait for our event source to connect
  await page.waitForResponse(eventSourceUrl);

  // fire the reload
  server.reload();

  // wait for the refresh
  const response = await page.waitForNavigation();

  // check it out
  assert.ok(response.ok());
  assert.equal(response.url(), `${url}/`);
});

it('should do an inline reload with CSS', async function () {
  // we need a new page
  const localPage = await browser.newPage();

  // create a temp dir
  const stylesDir = await fs.mkdtemp(join(tmpdir(), 'styles'));

  // prep the path to the styles
  const stylesPath = join(stylesDir, 'styles.css');

  // write out the css file
  await fs.writeFile(stylesPath, 'h1 { color: rgb(255, 0, 0); }');

  // create a new server to see the temp directory
  const localServer = create({
    dir: [join(__dirname, 'fixtures/styled'), stylesDir],
  });

  // start the server
  const { local } = await localServer.start();

  // get to the page
  await localPage.goto(local);

  // wait for the element to be available
  const el = await localPage.waitForSelector('h1');

  // get the initial color
  const startingColor = await localPage.evaluate(
    (el) => getComputedStyle(el).color,
    el
  );

  // confirm it matches
  assert.equal(startingColor, 'rgb(255, 0, 0)');

  // let's change the color
  await fs.writeFile(stylesPath, 'h1 { color: rgb(0, 0, 255); }');

  // tell the server to update
  localServer.reload('styles.css');

  // wait for our new styles to be requested and loaded
  const request = await localPage.waitForRequest(
    `${local}/styles.css?livereload=*`
  );
  await localPage.waitForResponse(request.url());

  // get the new color
  const endingColor = await localPage.evaluate(
    (el) => getComputedStyle(el).color,
    el
  );

  // confirm it changed
  assert.equal(endingColor, 'rgb(0, 0, 255)');

  // close the localPage
  await localPage.close();

  // close the local server
  await localServer.close();
});

it.run();


client/src/index.js
// packages
import { Reloader } from 'livereload-js/src/reloader';
import { Timer } from 'livereload-js/src/timer';

const prefix = '[MINI SYNC]';
const reloader = new Reloader(window, { log: () => {} }, Timer);
const reloadOptions = {
  liveCSS: true,
  liveImg: true,
};

function connect() {
  const source = new EventSource(
    `http://${window.location.hostname}:${window.location.port}/__mini_sync__`
  );

  source.addEventListener('open', () => {
    console.info('%s Development server has connected.', prefix);
  });

  source.addEventListener('error', (/** @type {ErrorEvent} */ error) => {
    const readyState = source.readyState;

    const isConnecting = readyState === EventSource.CONNECTING;
    const isClosed = readyState === EventSource.CLOSED;

    if (isConnecting || isClosed) {
      console.info('%s Lost connection. Trying to reconnect...', prefix);

      if (isClosed) {
        source.close();
        setTimeout(connect, 1e4);
      }
    } else {
      console.error(error);
    }
  });

  source.addEventListener('reload', (/** @type {MessageEvent} */ event) => {
    const data = JSON.parse(event.data);
    const file = data.file || '';

    if (file) {
      console.info('%s Reloading "%s".', prefix, file);
    } else {
      console.info('%s Reloading entire page.', prefix);
    }

    reloader.reload(file, reloadOptions);
  });
}

connect();


server.js
// native
const fs = require('fs');
const http = require('http');
const { join, normalize, resolve } = require('path');

// packages
const access = require('local-access');
const polka = require('polka');
const parse = require('@polka/url');
const sirv = require('sirv');

// package.json
const pkg = require('./package.json');

/**
 * Constant for repeated pinging.
 *
 * @private
 * @type {string}
 */
const ping = ':ping\n';

/**
 * A helper for sending SSE messages.
 *
 * @private
 * @param {string[]} messages A list of messages to send
 * @return {string}
 */
function writeMessage(messages) {
  return messages.join('\n') + '\n\n';
}

/**
 * Constant for reused Cache-Control header.
 *
 * @private
 * @type {string}
 */
const doNotCache = 'no-cache, no-store, must-revalidate';

/**
 * The contents of the compiled client-side code as a Buffer.
 * @private
 * @type {Buffer}
 */
const clientScript = fs.readFileSync(resolve(__dirname, pkg['umd:main']));

/**
 * The favicon.ico as a Buffer.
 * @private
 * @type {Buffer}
 */
const favicon = fs.readFileSync(resolve(__dirname, 'assets/favicon.ico'));

function htmlInjectionMiddleware(dirs) {
  const extensions = ['html', 'htm'];

  function getValue(x) {
    let y = x.indexOf('/', 1);
    return y > 1 ? x.substring(0, y) : x;
  }

  function getPossiblePaths(uri) {
    let i = 0;
    let x;
    let len = uri.length - 1;

    if (uri.charCodeAt(len) === 47) {
      uri = uri.substring(0, len);
    }

    const arr = [];
    let tmp = `${uri}/index`;

    for (; i < extensions.length; i++) {
      x = extensions[i] ? `.${extensions[i]}` : '';
      if (uri) arr.push(uri + x);
      arr.push(tmp + x);
    }

    return arr;
  }

  function queryLocalFile(dir, arr) {
    let i = 0;
    let abs, stats, name, headers;

    for (; i < arr.length; i++) {
      abs = normalize(join(dir, (name = arr[i])));

      if (abs.startsWith(dir) && fs.existsSync(abs)) {
        stats = fs.statSync(abs);
        if (stats.isDirectory()) continue;

        headers = {
          'Content-Type': 'text/html',
          'Last-Modified': stats.mtime.toUTCString(),
          'Cache-Control': doNotCache,
        };

        return { abs, stats, headers };
      }
    }
  }

  return function htmlInjection(req, res, next) {
    const info = parse(req);
    const paths = getPossiblePaths(getValue(info.pathname));

    for (let idx = 0; idx < dirs.length; idx++) {
      const dir = dirs[idx];

      const data = queryLocalFile(dir, paths);

      if (data) {
        const rendered = fs
          .readFileSync(data.abs, 'utf8')
          .replace(
            '</head>',
            '<script async src="/__mini_sync__/client.js"></script>\n</head>'
          );
        data.headers['Content-Length'] = Buffer.byteLength(rendered, 'utf8');
        res.writeHead(200, data.headers);
        res.end(rendered);
      }
    }

    next();
  };
}

/**
 * What's returned when the `create` function is called.
 *
 * @typedef {object} CreateReturn
 * @property {Function} close Stops the server if it is running
 * @property {Function} reload When called this function will reload any connected HTML documents, can accept the path to a file to target for reload
 * @property {Function} start When called the server will begin running
 */

/**
 * What's returned by the `start` function in a Promise.
 *
 * @typedef {object} StartReturn
 * @property {string} local The localhost URL for the static site
 * @property {string} network The local networked URL for the static site
 * @property {number} port The port the server ended up on
 */

/**
 * Creates a server on the preferred port and begins serving the provided
 * directories locally.
 *
 * @param {object} options
 * @param {string|string[]} [options.dir] The directory or list of directories to serve
 * @param {number} [options.port] The port to serve on
 * @param {boolean} [options.injectClientScript] If true, inject the client script into served HTML pages
 * @return {CreateReturn}
 * @example
 * const { create } = require('mini-sync');
 *
 * const server = create({ dir: 'app', port: 3000 });
 *
 * const { local } = await server.start();
 *
 * console.log(`Now serving at: ${local}`);
 *
 * // any time a file needs to change, use "reload"
 * server.reload('app.css');
 *
 * // reloads the whole page
 * server.reload();
 *
 * // close the server
 * await server.close();
 *
 */
function create({
  dir = process.cwd(),
  injectClientScript = true,
  port = 3000,
} = {}) {
  // create a raw instance of http.Server so we can hook into it
  const server = http.createServer();

  // a Set to track all the current client connections
  const clients = new Set();

  // create our polka server
  const app = polka({ server });

  app.get('/__mini_sync__', (req, res) => {
    //send headers for event-stream connection
    res.writeHead(200, {
      'Content-Type': 'text/event-stream',
      'Cache-Control': doNotCache,
      'Connection': 'keep-alive',
    });

    // send initial ping with retry command
    res.write('retry: 10000\n');

    // add client to Set
    clients.add(res);

    function disconnect() {
      // close the response
      res.end();

      // remove client from our Set
      clients.delete(res);
    }

    req.on('error', disconnect);
    res.on('error', disconnect);
    res.on('close', disconnect);
    res.on('finish', disconnect);
  });

  app.get('__mini_sync__/client.js', (_, res) => {
    // send headers for shipping down a JS file
    res.writeHead(200, {
      'Cache-Control': doNotCache,
      'Content-Length': clientScript.byteLength,
      'Content-Type': 'text/javascript',
    });

    // send the client-side script Buffer
    res.end(clientScript);
  });

  app.get('/favicon.ico', (_, res) => {
    // send headers for shipping down favicon
    res.writeHead(200, {
      'Cache-Control': doNotCache,
      'Content-Length': favicon.byteLength,
      'Content-Type': 'image/vnd.microsoft.icon',
    });

    res.end(favicon);
  });

  // make sure "serve" is an array
  const toWatch = Array.isArray(dir) ? dir : [dir];

  if (injectClientScript) {
    app.use(htmlInjectionMiddleware(toWatch));
  }

  // add each directory in "serve"
  for (let idx = 0; idx < toWatch.length; idx++) {
    const directory = toWatch[idx];

    app.use(sirv(directory, { dev: true }));
  }

  /**
   * Tells all connected clients to reload.
   *
   * @param {string} [file] The file to reload
   */
  function reload(file) {
    for (const client of clients) {
      client.write(
        writeMessage(['event: reload', `data: ${JSON.stringify({ file })}`])
      );
    }
  }

  function sendPing() {
    for (const client of clients) {
      client.write(ping);
    }
  }

  /**
   * Returns a promise once the server closes.
   *
   * @returns {Promise<void>}
   */
  function close() {
    return new Promise((resolve, reject) => {
      server.close((err) => {
        if (err) reject(err);

        resolve();
      });
    });
  }

  /**
   * Returns a promise once the server starts.
   *
   * @returns {Promise<StartReturn>}
   */
  function start() {
    return new Promise((resolve, reject) => {
      server.on('error', (e) => {
        if (e.code === 'EADDRINUSE') {
          setTimeout(() => {
            server.close();
            server.listen(++port);
          }, 100);
        } else {
          reject(e);
        }
      });

      let interval;

      server.on('listening', () => {
        // ping every 10 seconds
        interval = setInterval(sendPing, 10e3);

        // get paths to networks
        const { local, network } = access({ port });

        resolve({ local, network, port });
      });

      server.on('close', () => {
        if (interval) {
          clearInterval(interval);
          interval = null;
        }
      });

      app.listen(port);
    });
  }

  return { close, reload, start };
}

module.exports = { create };
